components/omega/doc/design/Metadata.md

```diff
@@ -0,0 +1,320 @@
+
+Internally, the metadata class will make use of the C++ 
+`std::map` that provides a useful capability for name, value
+pairs of each data type.
+
+### 4.1 Data types and parameters
+
+#### 4.1.1 Parameters 
+
+To enable general metadata associated with the code,
+simulation or file, we define standard names for these
+to be used as the "field" name during metadata definition.
+For code and simulation metadata, we will use:
+
+    static const std::string codeMeta = "code";
+    static const std::string simMeta  = "simulation";
+
+For file metadata (eg. time stamps, convention, etc.),
+we will use the stream name as the field name for metadata
+definition.
+
+All defined metadata (available fields) is kept public so
+that the module owning the data can define the metadata and the
+I/O routines can access all defined metadata.
+
+    static int nDefMeta; // number of fields defined
+    static std::vector<Metadata> definedMeta;
+
+Similarly, there will be a list of defined dimensions
+
+    static int nDefDims; // number of defined dimensions
+    static std::vector<MetadataDim> definedDims;
+
+#### 4.1.2 Class/structs/data types
+
+There are two classes associated with metadata. The first is
+the actual metadata class to hold the collection of metadata
+associated with a variable or general category (code, sim, file).
+This class is a collection of maps (name, value pairs) for 
+metadata of each supported data type. Note that there is none
+associated to the Real type since the type will need to
+correspond to the proper netCDF type. The Real type will
+be assigned the appropriate R4 or R8 by the compiler through
+the correct choice of aliased interfaces in the methods below.
+
+    class Metadata{
+
+       private:
+           std::string name; // field name to which metadata attached
+                             // use reserved names "code", "simulation"
+                             // or the I/O stream name for general
+                             // metadata associated with those
+
+           I4 numI4;   // number of I4 metadata entries
+           I4 numI8;   // number of I8 metadata entries
+           I4 numR4;   // number of R4 metadata entries
+           I4 numR8;   // number of R8 metadata entries
+           I4 numBool; // number of Boolean metadata entries
+           I4 numStr;  // number of string metadata entries
+
+           // Metadata name,value pairs for each type
+           std::map<std::string, I4> metaMapI4;
+           std::map<std::string, I8> metaMapI8;
+           std::map<std::string, R4> metaMapR4;
+           std::map<std::string, R8> metaMapR8;
+           std::map<std::string, bool> metaMapBool;
+           std::map<std::string, std::string> metaMapStr;
+
+           // Dimensions - only for defining fields/variables
+           I4 nDims;   // number of dimensions
+           std::vector<MetadataDim*> dims; // pointer to the defined
+                                           // dimension for each dim
```

Comment:
In MPAS-Ocean to support fill-value masking we had to add a pointer `mask` to `cellMask` or `edgeMask` (and should include other masks such as `landIceMask`). I don't know if you had another way around this in mind or if we need something similar here.

Reply:
At this level, the fillValue is only specifying the value to be used as a fill so that apps can check for that value in the field. The actual filling and masking will take place elsewhere.